pkg/internal/util/cli/logger.go

```diff
@@ -20,58 +20,105 @@ import (
 	"bytes"
 	"fmt"
 	"io"
+	"runtime"
+	"strings"
 	"sync"
 
 	"sigs.k8s.io/kind/pkg/log"
 )
 
 // Logger is the kind cli's log.Logger implementation
 type Logger struct {
-	Verbosity log.Level
-	io.Writer
-	writeMu sync.Mutex
+	writer     io.Writer
+	writerMu   sync.Mutex
+	verbosity  log.Level
+	bufferPool *bufferPool
 }
 
 var _ log.Logger = &Logger{}
 
 // NewLogger returns a new Logger with the given verbosity
 func NewLogger(writer io.Writer, verbosity log.Level) *Logger {
 	return &Logger{
-		Verbosity: verbosity,
-		Writer:    writer,
+		verbosity:  verbosity,
+		writer:     writer,
+		bufferPool: newBufferPool(),
 	}
 }
 
-func (l *Logger) Write(p []byte) (n int, err error) {
-	// TODO: line oriented instead?
-	// For now we make a single per-message write call from the rest of the logger
-	// intentionally to effectively do this one level up
-	l.writeMu.Lock()
-	defer l.writeMu.Unlock()
-	return l.Writer.Write(p)
+// synchronized write to the inner writer
+func (l *Logger) write(p []byte) (n int, err error) {
+	l.writerMu.Lock()
+	defer l.writerMu.Unlock()
+	return l.writer.Write(p)
 }
 
-// TODO: prefix log lines with metadata (log level? timestamp?)
-
-func (l *Logger) print(message string) {
-	buf := bytes.NewBufferString(message)
+// writeBuffer writes buf with write, ensuring there is a trailing newline
+func (l *Logger) writeBuffer(buf *bytes.Buffer) {
+	// ensure trailing newline
 	if buf.Bytes()[buf.Len()-1] != '\n' {
 		buf.WriteByte('\n')
 	}
 	// TODO: should we handle this somehow??
 	// Who logs for the logger? 🤔
-	_, _ = l.Write(buf.Bytes())
+	_, _ = l.write(buf.Bytes())
+}
+
+// print writes a simple string to the log writer
+func (l *Logger) print(message string) {
+	buf := bytes.NewBufferString(message)
+	l.writeBuffer(buf)
 }
 
+// printf is roughly fmt.Fprintf against the log writer
 func (l *Logger) printf(format string, args ...interface{}) {
-	var buf bytes.Buffer
-	fmt.Fprintf(&buf, format, args...)
-	if buf.Bytes()[buf.Len()-1] != '\n' {
-		buf.WriteByte('\n')
+	buf := l.bufferPool.Get()
+	fmt.Fprintf(buf, format, args...)
+	l.writeBuffer(buf)
+	l.bufferPool.Put(buf)
+}
+
+// addDebugHeader inserts the debug line header to buf
+func addDebugHeader(buf *bytes.Buffer) {
+	_, file, line, ok := runtime.Caller(3)
+	// lifted from klog
+	if !ok {
+		file = "???"
+		line = 1
+	} else {
+		if slash := strings.LastIndex(file, "/"); slash >= 0 {
+			path := file
+			file = path[slash+1:]
+			if dirsep := strings.LastIndex(path[:slash], "/"); dirsep >= 0 {
+				file = path[dirsep+1:]
+			}
+		}
 	}
-	// TODO: should we handle this somehow??
-	// Who logs for the logger? 🤔
-	_, _ = l.Write(buf.Bytes())
+	buf.Grow(len(file) + 11) // we know at least this many bytes are needed
+	buf.WriteString("DEBUG: ")
+	buf.WriteString(file)
+	buf.WriteByte(':')
+	fmt.Fprintf(buf, "%d", line)
```

Comment:
klog has a somewhat complex (and possibly incorrect for extremely large files?) optimization of this, it's difficult to do correctly and portably and cheaply by hand in go because this value is machine width :/

if we _Really_ care we can write a fancy on in a future commit. I don't think we do but it might be an interesting task ;-)